File: templates/api/go-gin/internal/handlers/posts.go
package handlers

import (
	"net/http"
	"strconv"

	"{{PROJECT_NAME}}/internal/middleware"
	"{{PROJECT_NAME}}/internal/models"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

type PostHandler struct {
	db *gorm.DB
}

func NewPostHandler(db *gorm.DB) *PostHandler {
	return &PostHandler{db: db}
}

// GetPosts godoc
// @Summary Get all posts
// @Description Get list of all posts with user information
// @Tags posts
// @Produce json
// @Success 200 {array} models.PostResponse
// @Router /posts [get]
func (h *PostHandler) GetPosts(c *gin.Context) {
	var posts []models.Post
	if err := h.db.Preload("User").Find(&posts).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch posts"})
		return
	}

	var response []models.PostResponse
	for _, post := range posts {
		response = append(response, post.ToResponse())
	}

	c.JSON(http.StatusOK, response)
}

// GetPost godoc
// @Summary Get post by ID
// @Description Get a single post by its ID
// @Tags posts
// @Produce json
// @Param id path int true "Post ID"
// @Success 200 {object} models.PostResponse
// @Failure 404 {object} map[string]string
// @Router /posts/{id} [get]
func (h *PostHandler) GetPost(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid post ID"})
		return
	}

	var post models.Post
	if err := h.db.Preload("User").First(&post, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Post not found"})
		return
	}

	c.JSON(http.StatusOK, post.ToResponse())
}

// CreatePost godoc
// @Summary Create a new post
// @Description Create a new post (requires authentication)
// @Tags posts
// @Accept json
// @Produce json
// @Security ApiKeyAuth
// @Param post body models.PostCreate true "Post data"
// @Success 201 {object} models.PostResponse
// @Failure 400 {object} map[string]string
// @Failure 401 {object} map[string]string
// @Router /posts [post]
func (h *PostHandler) CreatePost(c *gin.Context) {
	userID, ok := middleware.GetUserID(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
		return
	}

	var req models.PostCreate
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	post := models.Post{
		Title:   req.Title,
		Content: req.Content,
		UserID:  userID,
	}

	if err := h.db.Create(&post).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create post"})
		return
	}

	if err := h.db.Preload("User").First(&post, post.ID).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch created post"})
		return
	}

	c.JSON(http.StatusCreated, post.ToResponse())
}

// UpdatePost godoc
// @Summary Update a post
// @Description Update a post (requires authentication and ownership)
// @Tags posts
// @Accept json
// @Produce json
// @Security ApiKeyAuth
// @Param id path int true "Post ID"
// @Param post body models.PostUpdate true "Updated post data"
// @Success 200 {object} models.PostResponse
// @Failure 400 {object} map[string]string
// @Failure 401 {object} map[string]string
// @Failure 403 {object} map[string]string
// @Failure 404 {object} map[string]string
// @Router /posts/{id} [put]
func (h *PostHandler) UpdatePost(c *gin.Context) {
	userID, ok := middleware.GetUserID(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
		return
	}

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid post ID"})
		return
	}

	var post models.Post
	if err := h.db.First(&post, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Post not found"})
		return
	}

	if post.UserID != userID {
		c.JSON(http.StatusForbidden, gin.H{"error": "You can only update your own posts"})
		return
	}

	var req models.PostUpdate
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if req.Title != "" {
		post.Title = req.Title
	}
	if req.Content != "" {
		post.Content = req.Content
	}

	if err := h.db.Save(&post).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update post"})
		return
	}

	if err := h.db.Preload("User").First(&post, post.ID).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch updated post"})
		return
	}

	c.JSON(http.StatusOK, post.ToResponse())
}

// DeletePost godoc
// @Summary Delete a post
// @Description Delete a post (requires authentication and ownership)
// @Tags posts
// @Security ApiKeyAuth
// @Param id path int true "Post ID"
// @Success 204
// @Failure 400 {object} map[string]string
// @Failure 401 {object} map[string]string
// @Failure 403 {object} map[string]string
// @Failure 404 {object} map[string]string
// @Router /posts/{id} [delete]
func (h *PostHandler) DeletePost(c *gin.Context) {
	userID, ok := middleware.GetUserID(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
		return
	}

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid post ID"})
		return
	}

	var post models.Post
	if err := h.db.First(&post, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Post not found"})
		return
	}

	if post.UserID != userID {
		c.JSON(http.StatusForbidden, gin.H{"error": "You can only delete your own posts"})
		return
	}

	if err := h.db.Delete(&post).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete post"})
		return
	}

	c.Status(http.StatusNoContent)
}

File: templates/api/rust-actix/src/handlers/posts.rs
use actix_web::{web, HttpRequest, HttpResponse, Result, Scope};
use uuid::Uuid;
use validator::Validate;

use crate::{
    database::Database,
    middleware::AuthMiddleware,
    models::{CreatePostRequest, Post, PostResponse, PostWithUser, UpdatePostRequest},
};

pub fn config() -> Scope {
    web::scope("/posts")
        .route("", web::get().to(get_posts))
        .route("/{id}", web::get().to(get_post))
        .service(
            web::scope("")
                .wrap(AuthMiddleware)
                .route("", web::post().to(create_post))
                .route("/{id}", web::put().to(update_post))
                .route("/{id}", web::delete().to(delete_post))
        )
}

pub async fn get_posts(db: web::Data<Database>) -> Result<HttpResponse> {
    let pool = db.get_pool();

    let posts = sqlx::query_as!(
        PostWithUser,
        r#"
        SELECT
            p.id, p.title, p.content, p.user_id, p.created_at, p.updated_at,
            u.email as user_email, u.first_name as user_first_name,
            u.last_name as user_last_name, u.created_at as user_created_at,
            u.updated_at as user_updated_at
        FROM posts p
        JOIN users u ON p.user_id = u.id
        ORDER BY p.created_at DESC
        "#
    )
    .fetch_all(pool)
    .await;

    match posts {
        Ok(posts) => {
            let responses: Vec<PostResponse> = posts.into_iter().map(PostResponse::from).collect();
            Ok(HttpResponse::Ok().json(responses))
        }
        Err(e) => {
            log::error!("Database error: {:?}", e);
            Ok(HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Failed to fetch posts"
            })))
        }
    }
}

pub async fn get_post(
    path: web::Path<Uuid>,
    db: web::Data<Database>,
) -> Result<HttpResponse> {
    let post_id = path.into_inner();
    let pool = db.get_pool();

    let post = sqlx::query_as!(
        PostWithUser,
        r#"
        SELECT
            p.id, p.title, p.content, p.user_id, p.created_at, p.updated_at,
            u.email as user_email, u.first_name as user_first_name,
            u.last_name as user_last_name, u.created_at as user_created_at,
            u.updated_at as user_updated_at
        FROM posts p
        JOIN users u ON p.user_id = u.id
        WHERE p.id = $1
        "#,
        post_id
    )
    .fetch_optional(pool)
    .await;

    match post {
        Ok(Some(post)) => Ok(HttpResponse::Ok().json(PostResponse::from(post))),
        Ok(None) => Ok(HttpResponse::NotFound().json(serde_json::json!({
            "error": "Post not found"
        }))),
        Err(e) => {
            log::error!("Database error: {:?}", e);
            Ok(HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Database error"
            })))
        }
    }
}

pub async fn create_post(
    req: HttpRequest,
    body: web::Json<CreatePostRequest>,
    db: web::Data<Database>,
) -> Result<HttpResponse> {
    if let Err(errors) = body.validate() {
        return Ok(HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Validation failed",
            "details": errors
        })));
    }

    let user_id = match req.extensions().get::<Uuid>() {
        Some(id) => *id,
        None => {
            return Ok(HttpResponse::Unauthorized().json(serde_json::json!({
                "error": "Unauthorized"
            })));
        }
    };

    let pool = db.get_pool();

    let post = sqlx::query_as!(
        Post,
        r#"
        INSERT INTO posts (title, content, user_id)
        VALUES ($1, $2, $3)
        RETURNING *
        "#,
        body.title,
        body.content,
        user_id
    )
    .fetch_one(pool)
    .await;

    match post {
        Ok(post) => {
            // Fetch the post with user information
            let post_with_user = sqlx::query_as!(
                PostWithUser,
                r#"
                SELECT
                    p.id, p.title, p.content, p.user_id, p.created_at, p.updated_at,
                    u.email as user_email, u.first_name as user_first_name,
                    u.last_name as user_last_name, u.created_at as user_created_at,
                    u.updated_at as user_updated_at
                FROM posts p
                JOIN users u ON p.user_id = u.id
                WHERE p.id = $1
                "#,
                post.id
            )
            .fetch_one(pool)
            .await;

            match post_with_user {
                Ok(post_with_user) => Ok(HttpResponse::Created().json(PostResponse::from(post_with_user))),
                Err(e) => {
                    log::error!("Database error fetching created post: {:?}", e);
                    Ok(HttpResponse::Created().json(PostResponse::from(post)))
                }
            }
        }
        Err(e) => {
            log::error!("Failed to create post: {:?}", e);
            Ok(HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Failed to create post"
            })))
        }
    }
}

pub async fn update_post(
    req: HttpRequest,
    path: web::Path<Uuid>,
    body: web::Json<UpdatePostRequest>,
    db: web::Data<Database>,
) -> Result<HttpResponse> {
    if let Err(errors) = body.validate() {
        return Ok(HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Validation failed",
            "details": errors
        })));
    }

    let user_id = match req.extensions().get::<Uuid>() {
        Some(id) => *id,
        None => {
            return Ok(HttpResponse::Unauthorized().json(serde_json::json!({
                "error": "Unauthorized"
            })));
        }
    };

    let post_id = path.into_inner();
    let pool = db.get_pool();

    // Check if post exists and user owns it
    let existing_post = sqlx::query_as!(
        Post,
        "SELECT * FROM posts WHERE id = $1",
        post_id
    )
    .fetch_optional(pool)
    .await;

    let existing_post = match existing_post {
        Ok(Some(post)) => post,
        Ok(None) => {
            return Ok(HttpResponse::NotFound().json(serde_json::json!({
                "error": "Post not found"
            })));
        }
        Err(e) => {
            log::error!("Database error: {:?}", e);
            return Ok(HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Database error"
            })));
        }
    };

    if existing_post.user_id != user_id {
        return Ok(HttpResponse::Forbidden().json(serde_json::json!({
            "error": "You can only update your own posts"
        })));
    }

    // Update post
    let updated_title = body.title.as_ref().unwrap_or(&existing_post.title);
    let updated_content = body.content.as_ref().unwrap_or(&existing_post.content);

    let updated_post = sqlx::query_as!(
        Post,
        r#"
        UPDATE posts
        SET title = $1, content = $2, updated_at = NOW()
        WHERE id = $3
        RETURNING *
        "#,
        updated_title,
        updated_content,
        post_id
    )
    .fetch_one(pool)
    .await;

    match updated_post {
        Ok(post) => {
            // Fetch the post with user information
            let post_with_user = sqlx::query_as!(
                PostWithUser,
                r#"
                SELECT
                    p.id, p.title, p.content, p.user_id, p.created_at, p.updated_at,
                    u.email as user_email, u.first_name as user_first_name,
                    u.last_name as user_last_name, u.created_at as user_created_at,
                    u.updated_at as user_updated_at
                FROM posts p
                JOIN users u ON p.user_id = u.id
                WHERE p.id = $1
                "#,
                post.id
            )
            .fetch_one(pool)
            .await;

            match post_with_user {
                Ok(post_with_user) => Ok(HttpResponse::Ok().json(PostResponse::from(post_with_user))),
                Err(e) => {
                    log::error!("Database error fetching updated post: {:?}", e);
                    Ok(HttpResponse::Ok().json(PostResponse::from(post)))
                }
            }
        }
        Err(e) => {
            log::error!("Failed to update post: {:?}", e);
            Ok(HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Failed to update post"
            })))
        }
    }
}

pub async fn delete_post(
    req: HttpRequest,
    path: web::Path<Uuid>,
    db: web::Data<Database>,
) -> Result<HttpResponse> {
    let user_id = match req.extensions().get::<Uuid>() {
        Some(id) => *id,
        None => {
            return Ok(HttpResponse::Unauthorized().json(serde_json::json!({
                "error": "Unauthorized"
            })));
        }
    };

    let post_id = path.into_inner();
    let pool = db.get_pool();

    // Check if post exists and user owns it
    let existing_post = sqlx::query_as!(
        Post,
        "SELECT * FROM posts WHERE id = $1",
        post_id
    )
    .fetch_optional(pool)
    .await;

    let existing_post = match existing_post {
        Ok(Some(post)) => post,
        Ok(None) => {
            return Ok(HttpResponse::NotFound().json(serde_json::json!({
                "error": "Post not found"
            })));
        }
        Err(e) => {
            log::error!("Database error: {:?}", e);
            return Ok(HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Database error"
            })));
        }
    };

    if existing_post.user_id != user_id {
        return Ok(HttpResponse::Forbidden().json(serde_json::json!({
            "error": "You can only delete your own posts"
        })));
    }

    // Delete post
    let result = sqlx::query!(
        "DELETE FROM posts WHERE id = $1",
        post_id
    )
    .execute(pool)
    .await;

    match result {
        Ok(_) => Ok(HttpResponse::NoContent().finish()),
        Err(e) => {
            log::error!("Failed to delete post: {:?}", e);
            Ok(HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Failed to delete post"
            })))
        }
    }
}